tensorboard/pip_package/deterministic_tar_gz.py

```diff
@@ -0,0 +1,150 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+#
+# Much of this module is forked from Servo's `command_base.py`,
+# originally written by Anton Ovchinnikov and also provided under the
+# Apache-2.0 License:
+#
+# https://github.com/servo/servo/blob/d9fdf42bfe53dab08ba38fcdb349e84355f4cb3e/python/servo/command_base.py
+# https://github.com/servo/servo/pull/12244
+#
+"""Generate `.tar.gz` archives deterministically.
+
+See <https://reproducible-builds.org/docs/archives/>.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+import contextlib
+import functools
+import gzip
+import itertools
+import locale
+import os
+import shutil
+import tarfile
+import tempfile
+
+
+def main():
+  parser = argparse.ArgumentParser()
+  parser.add_argument(
+      "archive",
+      metavar="ARCHIVE",
+      help="name for the output file; should probably end in '.tar.gz'",
+  )
+  parser.add_argument(
+      "directory",
```

Comment:
Hmm, okay. It’s not obvious to me that this _decreases_ the complexity,
because you now need to consider filename collisions (and you also need
to think at least in passing about what happens if the newly required
shell glob in the genrule fails). If you prefer that complexity to the
(4-line) `os.walk`, fine by me; otherwise, I can revert.

Reply:
It seems less complex to me because it only supports a single level hierarchy, so it's really just about bundling several files into one, rather than an arbitrary directory structure.  And it eliminates the need not just to walk the tree but also the filename sorting, which in turn eliminates the need for locale setting.  The onus is on the caller to specify the right set of files, and in our case this is easy for the caller.